1-Draft/RFCNNNN-Dispose-Script-Block.md

```diff
@@ -0,0 +1,189 @@
+
+Functions in PowerShell are highly versatile, but don't always have the capabilities to manage resources correctly.
+For example, in .NET there are many types which implement `IDisposable`, advising that their `Dispose()` method should be used to clean them up when they're no longer needed.
+In addition to these, there are some types which manage connections to remote hosts or databases which commonly include a `Close()` method which is often to be used prior to calling `Dispose()`.
+Currently, PowerShell functions do not have an effective way of using these resources in the correct manner.
+
+This is further complicated by the fact that any terminating error in the pipeline completely bypasses the `end` named block.
+As such, it cannot be relied upon for cleanup / disposal steps.
+
+This includes any time a user includes the pipeline cmdlet `Select-Object` with the `-First` parameter.
+Once the desired number of objects has been reached, the cmdlet throws an internal and deliberately hidden exception to halt the pipeline.
+This is highly effective, but can lead to undesirable circumstances if essential cleanup steps are then skipped as a result.
+
+Compiled cmdlets do not suffer this issue, as they can simply implement `IDisposable` and place the necessary steps in a `Dispose()` method of their own.
+The pipeline processor will correctly handle **cmdlets** that require disposal steps, but currently not **functions**.
+
+## Motivation
+
+As a developer,<br />
+I want to be able to ensure I am not leaving resources or connections open when a user utilizes a function I have written.
+
+As an administrator,<br />
+I want to be able to ensure that any necessary logging or other necessary administrative tasks cannot be skipped accidentally or deliberately by a user invoking a function in a pipeline and using `Select-Object -First X` to bypass logging which would normally take place in the `end` block.
+
+## Specification
+
+A `dispose{}` block is required for these use cases.
+
+### Execution
+
+The `dispose{}` block will execute:
+
+1. After `end{}` during normal command operation, and
+2. During command disposal that occurs when a terminating exception is thrown during pipeline operations.
+    - If the dispose block has already been called (as it is after the `end{}` step), it will simply return immediately.
+
+### Behaviour
+
+The `dispose{}` block _does not_ guarantee output, but neither does it prevent it.
+
+- If output is received from the `dispose{}` block during normal command operation, it will be sent along the pipeline as per usual.
+- If output is received from the `dispose{}` block during a halting pipeline, it will be **discarded**.
+
+As a result, documentation should advise that `dispose{}` is _not intended_ for any standard output and the limitations of doing so.
+Additionally, all other PowerShell data streams are accessible from `dispose{}`, meaning there is no issue using any of the following streams from within `dispose{}`:
```

Comment:
This emphasizes my previous point. You're suggesting documentation advise against something that is being allowed for a new feature. If you prevent this in the first place, then documentation simply needs to indicate that standard output from a dispose block is automatically discarded. No need to allow it at all.

Reply:
If you can find a way to actually suppress standard output without also preventing use of the error stream for non-terminating errors, I'm all ears. 🙂

Reply:
It's code -- it's definitely possible. I just might need additional modification. I don't have time to dig into it right now, but I would personally list that as a requirement for implementation of this as a PR.

Reply:
From what I've seen thus far, that would require fundamental changes to how the PS pipelines themselves work, and while it may be a desirable end goal and accompanying feature, I don't think it makes sense to include here.

That may have impact further than just here, after all, as most other code in the pipeline handling doesn't appear to distinguish between error / output pipes currently. 

I think the time investment necessary to make such a change and the potential implications for other parts of the codebase outweigh the need to prevent the possibility of output here.

Reply:
Added this idea to Alternate / Additional section at the bottom so it doesn't get buried and we can potentially revisit it during PR / at a later date if needed. :)